src/main/java/com/amplitude/HttpTransport.java

```diff
@@ -334,43 +323,71 @@ private void triggerEventCallbacks(List<Event> events, int status, String messag
   }
 
   private void addEventToBuffer(String userId, String deviceId, Event event) {
-    idToBuffer.compute(
-        userId,
-        (key, value) -> {
-          if (value == null) {
-            value = new ConcurrentHashMap<>();
-          }
-          value.compute(
-              deviceId,
-              (deviceKey, deviceValue) -> {
-                if (deviceValue == null) {
-                  deviceValue = new ConcurrentLinkedQueue<>();
-                }
-                deviceValue.add(event);
-                return deviceValue;
-              });
-          return value;
-        });
+    synchronized (bufferLock) {
+      if (!idToBuffer.containsKey(userId)) {
+        idToBuffer.put(userId, new HashMap<>());
+      }
+      if (!idToBuffer.get(userId).containsKey(deviceId)) {
+        idToBuffer.get(userId).put(deviceId, new ArrayList<>());
+      }
+      idToBuffer.get(userId).get(deviceId).add(event);
+    }
+    synchronized (counterLock) {
+      eventsInRetry++;
+    }
   }
 
-  private Queue<Event> getEventsFromBuffer(String userId, String deviceId) {
-    List<Queue<Event>> eventQueues = new ArrayList<>();
-    idToBuffer.compute(
-        userId,
-        (key, value) -> {
-          if (value == null) {
-            return null;
+  private List<Event> getEventsFromBuffer(String userId, String deviceId) {
+    synchronized (bufferLock) {
+      if (idToBuffer.containsKey(userId) && idToBuffer.get(userId).containsKey(deviceId)) {
+        List<Event> events = idToBuffer.get(userId).remove(deviceId);
+        if (idToBuffer.get(userId).isEmpty()) {
+          idToBuffer.remove(userId);
+        }
+        return events;
+      }
+    }
+    return null;
+  }
+
+  public boolean shouldWait(Event event) {
+      if (recordThrottledId && (throttledUserId.containsKey(event.userId) || throttledDeviceId.containsKey(event.deviceId))) {
+        return true;
+      }
+      return eventsInRetry >= Constants.MAX_CACHED_EVENTS;
+  }
+
+  public void setRecordThrottledId(boolean record) {
+    recordThrottledId = record;
+  }
+
+  class EventBufferComsumer implements Runnable {
+
+    @Override
+    public void run() {
+      while (true) {
```

Comment:
Would this thread block program exit? Or we need guidance for that?

Reply:
I'll add a boolean switch to control this thread and add a method in client class to shutdown client ( flush all events, wait retry to finish, wait the consumer thread to join)

Reply:
Changed back to trigger retry when response of first events sending attempt has certain status.